examples/systemcalls/systemcalls.c

```diff
@@ -80,20 +125,72 @@ bool do_exec_redirect(const char *outputfile, int count, ...)
         command[i] = va_arg(args, char *);
     }
     command[count] = NULL;
-    // this line is to avoid a compile warning before your implementation is complete
-    // and may be removed
-    command[count] = command[count];
+    va_end(args);
 
+    /* If there is no command then there is nothing to do */
+    if (command[0] == NULL) {
+        return false;
+    }
 
-/*
- * TODO
- *   Call execv, but first using https://stackoverflow.com/a/13784315/1446624 as a refernce,
- *   redirect standard out to a file specified by outputfile.
- *   The rest of the behaviour is same as do_exec()
- *
-*/
+    int wait_val = 0;
+    int ret_val = 0;
+    int errors = 0;
+    int fd = -1;
+    pid_t pid;
 
-    va_end(args);
+    /* Creates log file with 0644 permissions. */
+    fd = open(outputfile, (O_WRONLY | O_TRUNC | O_CREAT), \
+        (S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH));    
+
+    /* Output errors to stdout if log file creation failed */
+    if (fd < 0) {
+        perror("open() failed: ");
+    }
+
+    /* Create a child process */
+    pid = fork();
+
+    switch (pid) {
+        case -1:
+            /* Failed to create child process */
+            perror("fork() failed: ");
+            errors++;
+            break;
+        case 0:
+            /* Inside child process */
+            /* Redirect any output to same logfile as parent process */
+            if (dup2(fd, 1) < 0) {
+                perror("dup2() failed: ");
+                exit(EXIT_FAILURE);
+            }
+            close(fd);
+            ret_val = execv(command[0], command);
+            
+            /* If process failed to execute return false for error */
+            if (ret_val == -1) {
+                perror("execv() failed: ");
+                exit(EXIT_FAILURE);
+            }
+            break;
+        default:
+            /* Parent does cleanup of process, waits for child process to exit */
+            if (wait(&wait_val) == -1) {
+                perror("wait() failed: ");
+                errors++;
+            }
+    }
+
+    /* Done with log now, close file */
+    close(fd);
+
+    /* 
+    * If we made it here the return val is a valid wait status. Check it to 
+    * make sure that the command exited properly and did not return a non-zero 
+    * value.
+    */
+    if (WIFEXITED(wait_val) && (errors == 0)) {
+        return (WEXITSTATUS(wait_val) == EXIT_SUCCESS);
```

Comment:
good!
See discussion in class about avoiding duplication of logic on these two very similar functions.